### DISCUSSION TITLE
Validation Plugins kdbGet vs. kdbSet

### DISCUSSION THREAD

**Discussion by kodebach** [ElektraInitiative/libelektra#3670] [General]
Some plugins perform validations during `kdbGet` (`postgetstorage`) and `kdbSet` (`presetstorage`), while others do it only during `kdbSet`. This is inconsistent and we should find a standard way to do things.

Examples of validation during `kdbGet` and `kdbSet`: `type`, `macaddr`
Examples of validation only during `kdbSet`: `network`, `reference`

Performing validation only during `kdbSet` can lead to problems, if the storage file is modified directly. Also some plugins also perform transformations in `kdbGet`, in which case validation during `kdbGet` is a must. Therefore, I think the standard way should be to validate during `kdbGet` and during `kdbSet`.

There is also the question of what to do, if validation fails. During `kdbSet`, it is pretty obvious that there should be an error to prevent creating invalid files. However, in `kdbGet` it is not so clear. If we produce an error and make `kdbGet` fail, we can never use `kdbSet` to fix the problem. But if we don't produce an error (and e.g. only produce a warning), applications may accidentally load invalid configurations.

IMO the best solution would be to validate both during `kdbGet` and `kdbSet` and always produce errors, unless a global flag (passed via the global keyset) is active. If the global flag is set, validation errors during `kdbGet` only produce a warning, but in `kdbSet` we still get an error. This allows us to fix broken configurations, but by default applications will not use invalid configs.

**Comment by markus2330** [comment]
I think it is easier to turn validation warnings to errors (e.g. by high-level APIs), then to implement such a flag.

And a validation plugin could remove non-validating keys that are not required. The only remaining case is if a key is required and invalid we should actually make an error, even on `get`.
